jwst/outlier_detection/outlier_detection.py

```diff
@@ -26,9 +29,97 @@
 
 
 __all__ = ["OutlierDetection", "flag_cr", "abs_deriv"]
+    
+
+def modelwise_operation(func: Callable) -> Callable:
+    """
+    Decorator to facilitate computations over a ModelContainer
+    preserving the spatial axis but splitting up the time axis, 
+    i.e., one ImageModel at at time
+
+    Notes
+    -----
+    Input function first two parameters MUST be self, datamodel
+    """
+    @functools.wraps(func)
+    def wrapper(self, container: ModelContainer, *args, **kwargs):
+        if not isinstance(container, ModelContainer):
+            raise ValueError("Input must be a ModelContainer")
+        
+        # store original value of return_open, then turn it off for now
+        ropen_orig = container._return_open
+        container._return_open = False
+
+        result_list = []
+        for i, model in enumerate(container):
+
+            # perform the requested operation
+            model = datamodel_open(model)
+            result = func(self, model, *args, **kwargs)
+            result_list.append(result)
+
+            # clear memory
+            model.close()
+            del model
+        
+        # Reset ModelContainer attribute to original value
+        container._return_open = ropen_orig
+
+        return result_list
+    
+    return wrapper
+
+
+def timewise_operation(func: Callable, buffer: float = 1.0) -> Callable:
+    """
+    Decorator to facilitate computations over a ModelContainer
+    preserving the time axis but splitting up the spatial axis
+    using ModelContainer.get_sections()
+
+    Parameters
+    ----------
+    buffer : float
+        Buffer size in Mb for each section
+
+    output_shape : tuple
+        Shape of the output array. If None, same shape as input
+
+    Notes
+    -----
+    Input function first two parameters MUST be (self, section)
+    where section = (data, dq, (row1, row2))
+    """
+    @functools.wraps(func)
+    def wrapper(self, container: ModelContainer, *args, output_shape=None, **kwargs):
+        if not isinstance(container, ModelContainer):
+            raise ValueError("Input must be a ModelContainer")
+        
+        # store original value of return_open, then turn it off for now
+        ropen_orig = container._return_open
+        container._return_open = False
+        container.set_buffer(buffer)
+
+        if output_shape is None:
+            output_shape = (len(container), container[0].data.shape[0], container[0].data.shape[1])
```

Comment:
Since `_return_open` is set to `False` a few lines prior the container will return filenames (if it contains filenames), models (if it contains models) or either if it contains both (which can happen if the asn contains `group_id` or `tweakreg_catalog` for any member entries). If a filename is returned the output_shape determination here will fail as `container[0]` will return a filename. I think adding a `datamodel_open` (if the `container[0]` is a filename) would address this.